openwisp_notifications/notification_types.py

```diff
@@ -0,0 +1,55 @@
+from django.core.exceptions import ImproperlyConfigured
+from openwisp_notifications.settings import ADDITIONAL_NOTIFICATION_TYPES
+
+DEFAULT_NOTIFICATION_TYPES = {
+    'default': {'level': 'info', 'verb': 'default verb', 'name': 'Default Type'},
+}
+
+
+def deep_merge_dicts(dict1, dict2):
+    result = dict1.copy()
+    for key, value in dict2.items():
+        if isinstance(value, dict):
+            node = result.get(key, {})
+            result[key] = deep_merge_dicts(node, value)
+        else:
+            result[key] = value
+    return result
+
+
+def get_notification_types():
+    notification_types = deep_merge_dicts(
+        DEFAULT_NOTIFICATION_TYPES, ADDITIONAL_NOTIFICATION_TYPES
+    )
+    validate_notification_type(notification_types)
+    return notification_types
+
+
+def get_notification_types_choices():
+    notification_types = NOTIFICATION_TYPES
+    choices = []
+    for key in sorted(notification_types.keys()):
+        name = notification_types[key].get('name', key)
+        choices.append((key, name))
+    return choices
```

Comment:
I'm afraid we have to do something different. I realized this after I worked on the chart configuration of the monitoring system
I think this is executed as soon as this module is loaded, hence, the choices available to `Notification.type` will be only those ones who have been registered before.

If another module is loaded after openwisp-monitoring and registers a notification type, I think it will not produce any effect

I suggest to try this that I described above :point_up:.

The first solution that comes into my mind is that we have to add another data structure, `NOTIFICATION_CHOICES`, which is also manipulated by `register_notification_type` and `unregister_notifcation_type`, that way, the choices should be available also if the registration happens after openwisp-monitoring is loaded.

Reply:
I am afraid that I don't understand the point of defining choices properly. Is it there to maintain integrity of database and prevent user for defining any arbitrary notification type?

Reply:
I was looking for a way it can be implemented and found this in [Django Documentation](https://docs.djangoproject.com/en/3.0/ref/models/fields/#choices). 
> Note that choices can be any sequence object – not necessarily a list or tuple. This lets you construct choices dynamically. But if you find yourself hacking choices to be dynamic, you’re probably better off using a proper database table with a ForeignKey. **choices is meant for static data that doesn’t change much, if ever.**

> Note
> A new migration is created each time the order of choices changes.

Reply:
because notifications will be generated based on their notification types. At some point we should be able to run code like:

`Notification.generate(type='monitoring.device.up')`
`Notification.generate(type='monitoring.device.down')`
`Notification.generate(type='firmware_upgrder.device.upgrade_started')`
`Notification.generate(type='firmware_upgrder.device.upgrade_completed')`
`Notification.generate(type='firmware_upgrder.device.upgrade_failed')`

In order to do this, we need to be able to have other modules extend the types.

We cannot go for the ForeignKey option: that would create migrations in this module, which makes everything incredibly painful. If a user is developing a custom openwisp derivative and starts introducing new notification types, migrations will be created here, then one day we'll change some model detail here and we'll generate other migrations and make their migrations conflict with ours.

In general the stuff that the user can manipulate should go in the database, but in this case the end-users using the OpenWISP application or its derivatives won't have to alter notification types, only we programmers do, we will add new notification types based on the requirements we get from companies and organizations using openwisp.

Reply:
Thanks for clearing out my confusion. :smile: 
I hope `generate` method is used for sake of example, since `notify` signal is already handling this. :grimacing:

Reply:
yes is an example :+1: